### PR TITLE
Fix SnackBar assertions when configured with theme

### LINKED ISSUE #84935: `SnackBar` constructor assertions don't take into account `SnackBarThemeData.behavior`
There is a way to set snackbar behavior via `SnackBarThemeData`, but the margin can only be set via the `SnackBar`'s constructor. Due to this, when the margin is specified and the `behavior` is set to `SnackBarBehavior.floating` via the theme, the latter must be duplicated in the constructor call cause otherwise the following assertion throws.

https://github.com/flutter/flutter/blob/25ba467b5a9bf1a3a9c84ca70d7c807ac4da5200/packages/flutter/lib/src/material/snack_bar.dart#L259

The same is true for the `width` property. 
https://github.com/flutter/flutter/blob/25ba467b5a9bf1a3a9c84ca70d7c807ac4da5200/packages/flutter/lib/src/material/snack_bar.dart#L263

## Steps to Reproduce
Run the following code

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.light().copyWith(
        snackBarTheme: SnackBarThemeData(
          behavior: SnackBarBehavior.floating, // <--
        ),
      ),
      debugShowCheckedModeBanner: false,
      home: ScaffoldMessenger(
        child: Scaffold(
          body: Center(
            child: MyWidget(),
          ),
        ),
      ),
    );
  }
}

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return TextButton(
      onPressed: () {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(
            content: Text('Test'),
            // behavior: SnackBarBehavior.floating, // assertion error without this
            margin: const EdgeInsets.all(16),
          ),
        );
      },
      child: Text('Show'),
    );
  }
}
```

**Expected results:**
No errors since the behavior is set via the theme

**Actual results:** 
```
Margin can only be used with floating behavior
'package:flutter/src/material/snack_bar.dart':
Failed assertion: line 259 pos 10: 'margin == null || behavior == SnackBarBehavior.floating'
```

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel master, 2.3.0-17.0.pre.411, on Microsoft Windows [Version 10.0.19041.1052], locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Chrome - develop for the web
[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.8.2)
[✓] Android Studio (version 4.2.0)
[✓] VS Code (version 1.52.1)
[✓] Connected device (4 available)

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by Piinks**
This moves the SnackBar assertions regarding width/margin and behavior to account for when the behavior is set using the theme.

Fixes https://github.com/flutter/flutter/issues/84935

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Piinks** [comment]
@HansMuller all cleaned up, PTAL? :)

File: packages/flutter/lib/src/material/snack_bar.dart [hunk 1]
```diff
@@ -485,13 +485,17 @@ class _SnackBarState extends State<SnackBar> {
...
       // assert that our other properties are configured properly.
       if (snackBarBehavior != SnackBarBehavior.floating) {
-        final String setter = widget.behavior != null
-            ? 'in the SnackBar constructor'
-            : snackBarTheme.behavior != null
-              ? 'by the inherited SnackBarThemeData'
-              : 'by default';
-        final String messageStub = 'can only be used with floating behavior. '
-          'SnackBarBehavior.fixed was set $setter.';
+        late final String messageStub;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/snack_bar.dart:488]: Forgive me. 
```dart
    assert((){
      // Whether the behavior is set through the constructor or the theme,
      // assert that our other properties are configured properly.
      if (snackBarBehavior != SnackBarBehavior.floating) {
        String message(parameter) {
          final String prefix  = '$parameter can only be used with floating behavior. ';
          if (widget.behavior != null) {
            return '$prefix SnackBarBehavior.fixed was set in the SnackBar constructor.';
          } else if (snackBarTheme.behavior != null) {
            return '$prefix SnackBarBehavior.fixed was set by the inherited SnackBarThemeData.';
          } else {
            return '$prefix SnackBarBehavior.fixed was set by default.';
          }
        }
        assert(widget.margin == null, message('Margin'));
        assert(widget.width == null, message('Width'));
      }
      return true;
    }());
```

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/snack_bar.dart [hunk 2]
```diff
@@ -489,6 +481,26 @@ class _SnackBarState extends State<SnackBar> {
 
     final TextStyle? contentTextStyle = snackBarTheme.contentTextStyle ?? ThemeData(brightness: brightness).textTheme.subtitle1;
     final SnackBarBehavior snackBarBehavior = widget.behavior ?? snackBarTheme.behavior ?? SnackBarBehavior.fixed;
+    assert((){
+      // Whether the behavior is set through the constructor or the theme,
+      // assert that our other properties are configured properly.
+      if (snackBarBehavior != SnackBarBehavior.floating) {
+        String message(String parameter) {
+          final String prefix = '$parameter can only be used with floating behavior. ';
```
Comment (Piinks) [inline packages/flutter/lib/src/material/snack_bar.dart:489]:
```suggestion
          final String prefix = '$parameter can only be used with floating behavior.';
```

**Pull request merged**